File: test/ModularPipelines.UnitTests/Dependencies/SingleTypeParameterGetModuleTests.cs
using ModularPipelines.Context;
using ModularPipelines.Exceptions;
using ModularPipelines.Models;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;
using Status = ModularPipelines.Enums.Status;

namespace ModularPipelines.UnitTests.Dependencies;

/// <summary>
/// Tests for the simplified single-type-parameter GetModule API.
/// </summary>
public class SingleTypeParameterGetModuleTests : TestBase
{
    /// <summary>
    /// A simple module that returns a string value.
    /// </summary>
    private class StringModule : Module<string>
    {
        protected internal override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "Hello from StringModule";
        }
    }

    /// <summary>
    /// A module that returns a complex type.
    /// </summary>
    private class ComplexResultModule : Module<ComplexResult>
    {
        protected internal override async Task<ComplexResult?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return new ComplexResult { Id = 42, Name = "Test" };
        }
    }

    /// <summary>
    /// A complex result type for testing type inference.
    /// </summary>
    private record ComplexResult
    {
        public int Id { get; init; }
        public string? Name { get; init; }
    }

    /// <summary>
    /// A module that uses the new single-type-parameter GetModule API.
    /// </summary>
    [ModularPipelines.Attributes.DependsOn<StringModule>]
    private class ConsumerModule : Module<string>
    {
        protected internal override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            // Use the new single-type-parameter API
            var result = await context.GetModule<StringModule>();

            // Verify the result is properly typed (ModuleResult<string?>)
            if (result.IsSuccess)
            {
                return result.ValueOrDefault;
            }

            return "failed";
        }
    }

    /// <summary>
    /// A module that uses GetModule to access a complex result type.
    /// </summary>
    [ModularPipelines.Attributes.DependsOn<ComplexResultModule>]
    private class ComplexConsumerModule : Module<int>
    {
        protected internal override async Task<int> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            // Type inference should work: result is ModuleResult<ComplexResult?>
            var result = await context.GetModule<ComplexResultModule>();

            if (result.IsSuccess && result.ValueOrDefault is not null)
            {
                return result.ValueOrDefault.Id;
            }

            return -1;
        }
    }

    /// <summary>
    /// A module that uses GetModuleIfRegistered with single type parameter.
    /// </summary>
    [ModularPipelines.Attributes.DependsOn<StringModule>(Optional = true)]
    private class OptionalConsumerModule : Module<string>
    {
        protected internal override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            // Use the new single-type-parameter GetModuleIfRegistered API
            var module = context.GetModuleIfRegistered<StringModule>();

            if (module is not null)
            {
                var result = await module;
                return result.ValueOrDefault ?? "default";
            }

            return "not registered";
        }
    }

    /// <summary>
    /// A module that tries to get itself (should throw).
    /// </summary>
    private class SelfReferencingModule : Module<bool>
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            // This should throw ModuleReferencingSelfException
            _ = context.GetModule<SelfReferencingModule>();
            await Task.Yield();
            return true;
        }
    }

    /// <summary>
    /// A module that tries to get an unregistered module (should throw).
    /// </summary>
    private class UnregisteredConsumerModule : Module<bool>
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            // StringModule is not registered, should throw
            _ = context.GetModule<StringModule>();
            await Task.Yield();
            return true;
        }
    }

    [Test]
    public async Task GetModule_SingleTypeParameter_ReturnsCorrectlyTypedResult()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<StringModule>()
            .AddModule<ConsumerModule>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task GetModule_SingleTypeParameter_WithComplexType_InfersTypeCorrectly()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<ComplexResultModule>()
            .AddModule<ComplexConsumerModule>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task GetModuleIfRegistered_SingleTypeParameter_ReturnsModule_WhenRegistered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<StringModule>()
            .AddModule<OptionalConsumerModule>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task GetModuleIfRegistered_SingleTypeParameter_ReturnsNull_WhenNotRegistered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<OptionalConsumerModule>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task GetModule_SingleTypeParameter_ThrowsModuleReferencingSelfException()
    {
        var exception = await Assert.ThrowsAsync<ModuleFailedException>(
            async () => await TestPipelineHostBuilder.Create()
                .AddModule<SelfReferencingModule>()
                .ExecutePipelineAsync());

        await Assert.That(exception.InnerException).IsTypeOf<ModuleReferencingSelfException>();
    }

    [Test]
    public async Task GetModule_SingleTypeParameter_ThrowsModuleNotRegisteredException()
    {
        var exception = await Assert.ThrowsAsync<ModuleFailedException>(
            async () => await TestPipelineHostBuilder.Create()
                .AddModule<UnregisteredConsumerModule>()
                .ExecutePipelineAsync());

        await Assert.That(exception.InnerException).IsTypeOf<ModuleNotRegisteredException>();
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/SingleTypeParameterGetModuleTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open System.Threading.Tasks
open ModularPipelines.Attributes
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Exceptions
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module SingleTypeParameterGetModuleTestModules =
    type ComplexResult = {
        Id: int
        Name: string
    }

    type StringModule() =
        inherit Module<string>()

        override _.ExecuteAsync(_, _) =
            task {
                do! Task.Yield()
                return "Hello from StringModule"
            }

    type ComplexResultModule() =
        inherit Module<ComplexResult>()

        override _.ExecuteAsync(_, _) =
            task {
                do! Task.Yield()
                return { Id = 42; Name = "Test" }
            }

    [<DependsOn(typeof<StringModule>)>]
    type ConsumerModule() =
        inherit Module<string>()

        override _.ExecuteAsync(context, _) =
            task {
                let! result = context.GetModule<StringModule>()

                if result.IsSuccess then
                    return result.ValueOrDefault
                else
                    return "failed"
            }

    [<DependsOn(typeof<ComplexResultModule>)>]
    type ComplexConsumerModule() =
        inherit Module<int>()

        override _.ExecuteAsync(context, _) =
            task {
                let! result = context.GetModule<ComplexResultModule>()

                if result.IsSuccess && not (isNull result.ValueOrDefault) then
                    return result.ValueOrDefault.Id
                else
                    return -1
            }

    [<ModularPipelines.Attributes.DependsOn(typeof<StringModule>, Optional = true)>]
    type OptionalConsumerModule() =
        inherit Module<string>()

        override _.ExecuteAsync(context, _) =
            task {
                let moduleInstance = context.GetModuleIfRegistered<StringModule>()

                if isNull moduleInstance then
                    return "not registered"
                else
                    let! result = moduleInstance
                    return if isNull result.ValueOrDefault then "default" else result.ValueOrDefault
            }

    type SelfReferencingModule() =
        inherit Module<bool>()

        override _.ExecuteAsync(context, _) =
            task {
                let _ = context.GetModule<SelfReferencingModule>()
                do! Task.Yield()
                return true
            }

    type UnregisteredConsumerModule() =
        inherit Module<bool>()

        override _.ExecuteAsync(context, _) =
            task {
                let _ = context.GetModule<StringModule>()
                do! Task.Yield()
                return true
            }

type SingleTypeParameterGetModuleTests() =
    inherit TestBase()

    [<Test>]
    member _.GetModule_SingleTypeParameter_ReturnsCorrectlyTypedResult() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<SingleTypeParameterGetModuleTestModules.StringModule>()
                .AddModule<SingleTypeParameterGetModuleTestModules.ConsumerModule>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.GetModule_SingleTypeParameter_WithComplexType_InfersTypeCorrectly() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<SingleTypeParameterGetModuleTestModules.ComplexResultModule>()
                .AddModule<SingleTypeParameterGetModuleTestModules.ComplexConsumerModule>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.GetModuleIfRegistered_SingleTypeParameter_ReturnsModule_WhenRegistered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<SingleTypeParameterGetModuleTestModules.StringModule>()
                .AddModule<SingleTypeParameterGetModuleTestModules.OptionalConsumerModule>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.GetModuleIfRegistered_SingleTypeParameter_ReturnsNull_WhenNotRegistered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<SingleTypeParameterGetModuleTestModules.OptionalConsumerModule>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.GetModule_SingleTypeParameter_ThrowsModuleReferencingSelfException() = async {
        let mutable innerException = null

        try
            let! _ =
                TestPipelineHostBuilder.Create()
                    .AddModule<SingleTypeParameterGetModuleTestModules.SelfReferencingModule>()
                    .ExecutePipelineAsync()
                |> Async.AwaitTask
            ()
        with :? ModuleFailedException as ex ->
            innerException <- ex.InnerException

        do! check(Assert.That(innerException).IsTypeOf<ModuleReferencingSelfException>())
    }

    [<Test>]
    member _.GetModule_SingleTypeParameter_ThrowsModuleNotRegisteredException() = async {
        let mutable innerException = null

        try
            let! _ =
                TestPipelineHostBuilder.Create()
                    .AddModule<SingleTypeParameterGetModuleTestModules.UnregisteredConsumerModule>()
                    .ExecutePipelineAsync()
                |> Async.AwaitTask
            ()
        with :? ModuleFailedException as ex ->
            innerException <- ex.InnerException

        do! check(Assert.That(innerException).IsTypeOf<ModuleNotRegisteredException>())
    }
